### ISSUE TITLE
ENH: iterative clip using spatial index

### ISSUE THREAD

**Issue by knaaptime** [geopandas/geopandas#1803]
I'm using a geodataframe of ~24k polygons to clip a smaller geodataframe of ~400 polygons (the full example is [here](https://github.com/pysal/tobler/blob/master/notebooks/binary_dasymetric.ipynb)). The standard `clip` function isnt viable for this problem, taking many hours to complete so i wrote a simple function that uses the spatial index to iterate through the features clipping subcomponents before recombining them at the end

```python
def iterclip(source, mask):
    tree = mask.sindex
    features = []
    for i, row in source.iterrows():
        query = tree.query(source.geometry.iloc[i], predicate='intersects').tolist()
        submask = mask[mask.index.isin(query)]
        single = gpd.GeoDataFrame(row.to_frame().T, crs=source.crs)
        features.append(gpd.clip(single, submask))
    frame = gpd.GeoDataFrame(pd.concat(features), crs=source.crs)
    return frame
```

That handles my problem nicely so I'd be happy to open a PR if it's of interest here

**Comment by martinfleis** [comment]
This an interesting case, because we find here that the main bottleneck is probably `unary_union` of those 24k polygons. Also, this case makes no use of spatial index due to a large overlap. The case above can be likely further sped up by `query_bulk` instead of individual `query` calls within the loop and maybe `itertuples`/`apply`?

I am just thinking how to incorporate it best to the current clip because there are cases where the existing function is faster than this, e.g. if you clipped those 24k by those 400. So, whether `clip` should be smart about it and figure out the best approach, having a way to control it via keyword as an independent function as you have here. In my view, it is GeoPandas job to figure out how to clip stuff efficiently and the user should not be bothered about the actual implementation, trying to understand which of the options is better. I guess we will need to understand the performance effect more before making the decision, to know all the implications.

That said, this is certainly a super useful speedup which should be available.

**Comment by knaaptime** [comment]
yeah, i figured this is a somewhat special case, given the size of the problem, but also not terribly uncommon. I defaulted to an issue rather than a PR so it would be easier to work through the caveats you raise (and wasnt sure if you might prefer, e.g. a keyword). For small problems that can leverage the spatial index already, the breaking apart and iterating probably adds more overhead, so not sure what makes sense as the default approach. I agree ideally this is all figured out under the hood, but i wasnt sure how to go about determining those heuristics 

i'm happy to do whatever i can to help move this forward, but not 100% on next steps?

**Comment by martinfleis** [comment]
I think that the first step should be to understand in which cases this is faster and in which slower, so we can build the heuristics and understand it a bit more. Once that is clear, we can make a PR.

However, to do that properly we should try to speed this up (see my comment above) so we have a relevant code for benchmarks.

**Comment by jorisvandenbossche** [comment]
Interesting problem! One idea that I had is: we basically already have a kind of clip that doesn't first "union" your mask (which is the problem here, I think) and uses a spatial index, which is the overlay function with intersection. 

The main difference is that this not only clips your original geometries but also might cut it in many pieces if your mask consists of small geometries. 

But if you then union those afterwards (dissolve), then you might end up with the result you want.

I checked it with your example notebook:

```python
# dasy = gpd.clip(rside_census, urban_rside)

dasy2 = gpd.overlay(rside_census, urban_rside)
day2 = dasy2.dissolve("GEOID")
```

and where your `iterclip` function takes 2min20 (on my laptop, latest versions of all libs), the above takes 28 s (18 for the overlay, 10 for dissolving).

Of course, geopandas' overlay already makes use of `query_bulk`, so maybe rewriting the `iterclip` above to use that as well might give similar speed-ups. 
On the other hand, we are always going to union some geomtries: if not the result from overlay as I did, then the geometries to clip with before calculating the interesection. So it might not matter that much at which stage the geometries are unioned.

**Comment by jorisvandenbossche** [comment]
Based on a quick profile of `iterclip`, 90% of the time is still taken by the `unary_union` happening in `clip` for unioning the mask geometries. So although they have been limited to the minimum by the spatial query, this is still the bottleneck. Given that in this case the mask consists of much more complex geometries as the the left geometries (census), it might be that unioning the census geometries after clipping (so my overlay+dissolve approach) instead of unioning the mask geometries before clipping will be the fastest option.

**Comment by cheginit** [comment]
In a project that I am working on I had a similar issue. My solution is based on `query_bulk` and running it in parallel. Thought it might be useful to share it here:

```python
import cytoolz as tlz
import geopandas as gpd
import joblib
from dask.distributed import Client


def query_indicies(
    input_geo: gpd.GeoDataFrame,
    tree_geo: gpd.GeoDataFrame,
    predicate: str,
) -> Dict[int, List[int]]:
    """Find the indices of the input_geo that intersect with the tree_geo.

    Parameters
    ----------
    input_geo : gpd.GeoDataFrame
        The input geodataframe.
    tree_geo : gpd.GeoDataFrame
        The tree geodataframe.
    predicate : str, optional
        The predicate to use for the query operation, defaults to ``intesects``.

    Returns
    -------
    dict
        A dictionary of the indices of the input_geo that intersect with the tree_geo.
        Keys are the index of tree_geo and values are a list of the indices of input_geo
        that intersect with the tree_geo.
    """
    for crs in [input_geo.crs, tree_geo.crs]:
        if crs is not None and not crs.is_projected:
            raise ValueError("Input dataframes must be in a projected CRS.")

    if input_geo.crs != tree_geo.crs:
        raise ValueError("Both input dataframes must be in the same projected CRS.")

    input_idx, tree_idx = tree_geo.sindex.query_bulk(input_geo.geometry, predicate=predicate)
    return tlz.merge_with(list, ({t: i} for t, i in zip(tree_idx, input_idx)))


def clip_bulk(
    input_geo: gpd.GeoDataFrame,
    mask_geo: gpd.GeoDataFrame,
    partition: int = 800,
    local_cluster: bool = True,
) -> gpd.GeoDataFrame:
    """Clip the input geometry based the mask.

    Parameters
    ----------
    input_geo : gpd.GeoDataFrame
        The input geodataframe.
    mask_geo : gpd.GeoDataFrame
        The mask geodataframe.
    partition : int
        Number of partitions for parallelization.
    local_cluster : bool, optional
        Whether the cluster is local or not, default is True.

    Returns
    -------
    geopandas.GeoDataFrame
        Cliped input_geo.
    """
    merged_idx = query_indicies(mask_geo, input_geo, "intersects")

    def clip_coverage(geom: gpd.GeoDataFrame, mask: gpd.GeoDataFrame) -> gpd.GeoDataFrame:
        return gpd.clip(geom.geometry, mask.unary_union)

    def concat_idx(idx):
        return list(set(tlz.concat(merged_idx[i] for i in idx)))

    with Client(processes=not local_cluster), joblib.parallel_backend("dask"):
        out_geo = joblib.Parallel(verbose=0)(
            joblib.delayed(clip_coverage)(input_geo.iloc[list(idx)], mask_geo.iloc[concat_idx(idx)])
            for idx in tlz.partition_all(partition, list(merged_idx))
        )
    return gpd.GeoDataFrame(geometry=pd.concat(out_geo), crs=input_geo.crs)
```

**Comment by theroggy** [comment]
For people running in a performance issue with (very) large/complex input files this is yet another option. It needs (geopackage) files as input and writes the result to a file as well: [geofileops.clip](https://geofileops.readthedocs.io/en/stable/api/geofileops.clip.html#geofileops.clip)
